ReactAndroid/src/main/java/com/facebook/react/modules/settings/SettingsModule.java

```diff
@@ -0,0 +1,180 @@
+
+import java.util.HashMap;
+import java.util.Map;
+
+@ReactModule(name = SettingsModule.NAME)
+public class SettingsModule extends ReactContextBaseJavaModule implements SharedPreferences.OnSharedPreferenceChangeListener{
+
+  @SuppressWarnings("WeakerAccess")
+  static final String NAME = "SettingsManager";
+  static final String DEFAULT_FILE_NAME = "ReactNative";
+  static final String TAG = "SettingsMmodule";
+
+  private static boolean sSettingsModuleInstantiated = false;
+
+  static private String sFilename = "";
+  private Boolean ignoringUpdates = false;
+  private SharedPreferences mPreferences;
+
+  public SettingsModule(ReactApplicationContext context) {
+    super(context);
+
+  }
+
+  @Override
+  public String getName() {
+    return SettingsModule.NAME;
+  }
+
+  private SharedPreferences ensurePreferences() {
+    if (mPreferences == null) {
+      if (SettingsModule.sFilename.length() > 0) {
+        mPreferences = getReactApplicationContext().getSharedPreferences(SettingsModule.sFilename, Context.MODE_PRIVATE);
+      } else if (getCurrentActivity() != null) {
+        mPreferences = getCurrentActivity().getPreferences(Context.MODE_PRIVATE);
+      } else {
+        mPreferences = getReactApplicationContext().getSharedPreferences(DEFAULT_FILE_NAME, Context.MODE_PRIVATE);
+      }
+      sSettingsModuleInstantiated = true;
+      mPreferences.registerOnSharedPreferenceChangeListener(this);
+    }
+
+    return mPreferences;
+  }
+
+  /***
+   * Configure the Settings Module to use a SharedPreferences file with a specific name as its backing store.
+   * Useful if the Settings Module needs to use an existing SharedPreferences file or to ensure that Settings Modules
+   * belonging to different ReactPackages will all share the same SharedPreferences file for their backing store.
+   * <p></p>
+   * <b>Note:</b>this method may only be called before a Settings Module initialization (i.e. before a ReactMainPackage is instantiated)
+   * An app would typically invoke this function in the app's main activity before setting up the React Native environment
+   * @param filename the name of the SharedPreferences file to use or create for the backing store
+   * @throws IllegalStateException if called after instantiation of a Settings module
+   */
+  static public void setSharedPreferencesFilename(String filename) {
+    if (sSettingsModuleInstantiated) {
+      throw new IllegalStateException("Attempt to set Settings Module SharedPreferences file name after a Settings Module instantiation");
+    }
+    sFilename = filename;
+  }
+
+  @Override
+  public void onCatalystInstanceDestroy() {
+    ensurePreferences().unregisterOnSharedPreferenceChangeListener(this);
+  }
+
+  @ReactMethod
+  public void setValues(ReadableMap values) {
+    this.ignoringUpdates = true;
+
+    ReadableMapKeySetIterator iterator = values.keySetIterator();
+    SharedPreferences.Editor editor = mPreferences.edit();
```

Comment:
I hope I am not becoming annoying :D but why do you use `mPreferences` here, while you call `ensurePreferences()` in other methods? I believe there should be only one way of working with the preferences.